File: Projet/app/src/main/java/utbm/lepysidawy/code_p25/database/AppDatabase.java
package utbm.lepysidawy.code_p25.database;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import utbm.lepysidawy.code_p25.dao.ParticipateRaceDAO;
import utbm.lepysidawy.code_p25.dao.RaceDAO;
import utbm.lepysidawy.code_p25.dao.RunnerDAO;
import utbm.lepysidawy.code_p25.entity.ParticipateRace;
import utbm.lepysidawy.code_p25.entity.Race;
import utbm.lepysidawy.code_p25.entity.Runner;

/**
 * Singleton class for the database
 */
@Database(entities = {Runner.class, Race.class, ParticipateRace.class}, version = 6)
public abstract class AppDatabase extends RoomDatabase {

    public abstract RunnerDAO runnerDAO();
    public abstract RaceDAO raceDAO();
    public abstract ParticipateRaceDAO participateRaceDAO();

    private static AppDatabase instance;

    /**
     * Get the application database instance
     * @param context The application context
     * @return The database instance
     */
    public static synchronized AppDatabase getInstance(Context context){
        if(instance == null){
            //TODO: Remove allowMainThreadQueries, because it can freeze UI on big queries (but normally not on this project). Better to call queries outside of the UI thread.
            instance  = Room.databaseBuilder(context.getApplicationContext(), AppDatabase.class, "app_db").fallbackToDestructiveMigration().allowMainThreadQueries().build();
        }
        return instance;
    }
}

File: Projet/GestionCourse/Gestion-Course/app/src/main/java/com/example/gestion_course/AppDatabase.kt
package com.example.gestion_course

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.example.gestion_course.dao.*
import com.example.gestion_course.entities.Equipe
import com.example.gestion_course.entities.Etape
import com.example.gestion_course.entities.Participant
import com.example.gestion_course.entities.Temps
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

@Database(entities = arrayOf(Equipe::class, Participant::class, Etape::class, Temps::class), version = 6)
abstract class AppDatabase : RoomDatabase() {

    abstract fun equipeDao(): EquipeDao
    abstract fun participantDao(): ParticipantDao
    abstract fun etapeDao(): EtapeDao
    abstract fun tempsDao(): TempsDao

    companion object {
        private var instance: AppDatabase? = null

        @Synchronized
        fun getInstance(ctx: Context): AppDatabase {
            if (instance == null)
                instance = Room.databaseBuilder(
                    ctx.applicationContext, AppDatabase::class.java,
                    "app_database"
                )
                    .fallbackToDestructiveMigration()
                    .build()

            return instance!!

        }
    }

    fun clearTables() {
        GlobalScope.launch(Dispatchers.IO) {
            this@AppDatabase.clearAllTables()
        }
    }
}